### ISSUE TITLE
GRPC_ARG_KEEPALIVE_TIME_* on the C++ client does not work when streaming

### ISSUE THREAD

**Issue by forestphoenix** [grpc/grpc#14717]
### Should this be an issue in the gRPC issue tracker?
 yes
  
### What version of gRPC and what language are you using?

grpc 1.10 with C++
 
### What operating system (Linux, Windows, …) and version?

Ubuntu 16.04.3 LTS (Xenial Xerus)
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
 
gcc (Ubuntu 5.4.0-6ubuntu1~16.04.9) 5.4.0 20160609
 
### What did you do?
 
- I start a grpc server and a client on different machines
  Both server and client are started with
`GRPC_ARG_KEEPALIVE_TIME_MS = 10000` and
`GRPC_ARG_KEEPALIVE_TIMEOUT_MS = 10000`

- the client opens a bidirectional stream to the server (both sides write & read messages)

- I pull the cable between server & client

### What did you expect to see?
 
The `Read` and `Write` calls fail on both sides after 20 **seconds**.
 
### What did you see instead?

The `Read` and `Write` calls fail on both sides after 15 **minutes**.
 
### Anything else we should know about your project / environment?

When using plain RPC calls without streaming, the client does time out after 20 seconds.

**Comment by forestphoenix** [comment]
Source code to reproduce the issue can be found here:
https://gist.github.com/forestphoenix/cd9d0ae24cbb5c6218b4299f4df5521b

**Comment by yang-g** [comment]
Please try setting GRPC_ARG_KEEPALIVE_PERMIT_WITHOUT_CALLS to 1 at both client and server.
And set GRPC_ARG_HTTP2_MIN_RECV_PING_INTERVAL_WITHOUT_DATA_MS at server to a value smaller than 10000 and set GRPC_ARG_HTTP2_MIN_SENT_PING_INTERVAL_WITHOUT_DATA_MS at client to 10000.

**Comment by yang-g** [comment]
Feel free to reopen the issue if you still see problems. Thanks.

**Comment by forestphoenix** [comment]
I have added the arguments you suggested, but the both server and client still take 15 minutes to time out.

**Comment by forestphoenix** [comment]
@yang-g Given that the problem still exists, can you reopen the issue?

I since you (as contributor) closed the issue, I cannot reopen it.

**Comment by forestphoenix** [comment]
When I stop doing `Write` calls on the client, the client does time out after 20 seconds.

With `Write` calls, it takes 15 minutes.

Could that be a clue about what is going on here?

**Comment by yang-g** [comment]
I looked at your source code and it seems you are doing a non-stop sequence of writes. In this case, the write should detect the failure of the connection (rather than the keepalive), and the fact that  Write failed to come back with an error is the key problem here.

**Comment by yang-g** [comment]
@yashykt Could you take a look? Thanks.

**Comment by srini100** [comment]
@yang-g, why should write detect the failure of connection when TCP has not yet timed out (15 mins in this case)? May be the problem is that keepalive logic thinks the bytes are are being sent but they are just being buffered in TCP due to broken connection.

**Comment by yashykt** [comment]
I also agree with Yang's observation that the keepalive timer has no affect here since the writes are being done continuously at both the server and the client.

I took a look at the code, and the non-stop writes are probably resetting the keepalive timer each time a write is attempted which is intended afaik. Also, a write might not return an error immediately because the writes are probably getting buffered.

 If the timeout is a concern in this scenario, I would suggest disabling buffering in the WriteOptions, and waiting till the write is actually complete.

**Comment by srini100** [comment]
@yashykt, doesn't keepalive logic send a ping when no data frame has been received for GRPC_ARG_HTTP2_MIN_SENT_PING_INTERVAL_WITHOUT_DATA_MS duration?

**Comment by yashykt** [comment]
@srini100 it would send a ping but if we are already writing then the ping is not really required right? The write buffering over here is causing the issue imo.

**Comment by srini100** [comment]
In my mind the purpose of keepalive feature is to check the health of the connection (RTT, breakage etc.) without relying on lower layers. While we are sending data but not receiving anything, a ping should be sent after some duration to confirm that the connection is still alive. The ping response will confirm that connection is fine. I am assuming that ping is sent with tcp_nodelay option.

**Comment by yashykt** [comment]
That should be true. But the ping will only be sent if the timer expires
but if the timer is reset on a write then the ping will not be sent.

**Comment by yashykt** [comment]
@forestphoenix Also try setting GRPC_ARG_HTTP2_BDP_PROBE to 0

**Comment by yashykt** [comment]
Reopening it. Based on reading the relevant keepalive timer code we are not resetting the timer on writes or reads, but we do reset it on a bdp ping.

**Comment by srini100** [comment]
@yashykt, from the [design doc ](https://github.com/grpc/proposal/blob/master/A8-client-side-keepalive.md) "keepalive time is ideally measured from the time of the last byte read".

**Comment by forestphoenix** [comment]
@yashykt I tried setting the `GRPC_ARG_HTTP2_BDP_PROBE` to 0.
Doing that changed nothing.

I also did some investigating on my side:

Normally, a timer that calls `keepalive_watchdog_fired_locked` is activated when the ping is *processed* from the `ping_queue` of the transport and started.

In the scenario where I pull the cable and the client does `Write` calls, the ping never gets processed from the queue (from what I can tell, `maybe_initiate_ping` does not get called).

**Comment by yashykt** [comment]
@srini100 We don't seem to be following the ideal implementation.

**Comment by GupTk** [comment]
Hi everyone,
I encounter the same problem. 
Anything new from this matter?

Thanks

**Comment by yashykt** [comment]
@GupTk What does your scenario look like?

**Comment by GupTk** [comment]
> @GupTk What does your scenario look like?

Our Client (Windows) writes a stream of data to the server (an embedded Linux) and receives also streamed status information from the server.
We encountered the same problem when testing grpc’s failure modes for our application - e.g. when unplugging the network cable

**Comment by yashykt** [comment]
Sorry for the delay. I finally got around to debugging this. 
The issue comes up in the following scenario -
1) The client and server have an established channel between them.
2) The client sends a ping to the server. (This ping is not a keepalive ping.)
3) The network/connection goes down with the client still waiting to hear back the ACK on the ping.
4) The keepalive timer expires, but the keepalive ping is not sent, since there is already a ping in progress.
5) The keepalive ping would only be sent once we hear back the ACK which we aren't going to get. Also, the watchdog timer for the keepalive ping is only started once we send the keepalive ping. This leads to a hung state. 

The fix would be to start the watchdog timer for the keepalive ping if there is a ping already in progress. This would use the ping in progress as a keepalive ping too. This should avoid the condition where the client and server hang up. I'll send out a pull request for this.

**Comment by sreecha** [comment]
Thanks for the detailed comments Yash. This was very helpful in understanding how keepalive in grpc-core works!
